pkg/core/blockchain.go

```diff
@@ -613,18 +623,225 @@ func (bc *Blockchain) jumpToStateInternal(p uint32, stage stateJumpStage) error
+	atomic.StoreUint32(&bc.blockHeight, height)
+	atomic.StoreUint32(&bc.persistedHeight, height)
+
 	err = bc.initializeNativeCache(block.Index, bc.dao)
 	if err != nil {
 		return fmt.Errorf("failed to initialize natives cache: %w", err)
 	}
 
-	if err := bc.updateExtensibleWhitelist(p); err != nil {
+	if err := bc.updateExtensibleWhitelist(height); err != nil {
 		return fmt.Errorf("failed to update extensible whitelist: %w", err)
 	}
 
-	updateBlockHeightMetric(p)
+	updateBlockHeightMetric(height)
+	return nil
+}
 
-	bc.dao.Store.Delete(jumpStageKey)
+// Reset resets chain state to the specified height if possible. This method
+// performs rough DB changes and can be called on non-running Blockchain only.
+func (bc *Blockchain) Reset(height uint32) error {
+	bc.dao.PutStateSyncPoint(height)
+	return bc.resetStateInternal(height, none)
+}
+
+func (bc *Blockchain) resetStateInternal(height uint32, stage stateChangeStage) error {
+	currHeight := bc.BlockHeight()
+	if height > currHeight {
+		return fmt.Errorf("current block height is %d, can't reset state to height %d", currHeight, height)
+	}
+	if height == currHeight && stage == none {
+		bc.log.Info("chain is already at the proper state", zap.Uint32("height", height))
+		return nil
+	}
+	if bc.config.KeepOnlyLatestState {
+		return fmt.Errorf("KeepOnlyLatestState is enabled, state for height %d is outdated and removed from the storage", height)
+	}
+	if bc.config.RemoveUntraceableBlocks && currHeight >= bc.config.MaxTraceableBlocks {
+		return fmt.Errorf("RemoveUntraceableBlocks is enabled, a necessary batch of traceable blocks has already been removed")
+	}
+
+	// Retrieve necessary state before the DB modification.
+	hHeight := bc.HeaderHeight()
+	b, err := bc.GetBlock(bc.headerHashes[height])
+	if err != nil {
+		return fmt.Errorf("failed to retrieve block %d: %w", height, err)
+	}
+	sr, err := bc.stateRoot.GetStateRoot(height)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve stateroot for height %d: %w", height, err)
+	}
+	v := bc.dao.Version
+	cache := bc.dao // dao is MemCachedStore over DB, so use dao directly to persist cached changes right to the underlying DB
+
+	bc.log.Info("initialize state reset", zap.Uint32("height", height))
+	start := time.Now()
+	p := start
+
+	resetStageKey := []byte{byte(storage.SYSStateChangeStage)}
+	switch stage {
+	case none:
+		cache.Store.Put(resetStageKey, []byte{stateResetBit | byte(stateJumpStarted)})
+		fallthrough
+	case stateJumpStarted:
+		// Remove headers/blocks/transactions/aers from currHeight down to height (not including height itself).
+		for i := height + 1; i <= hHeight; i++ {
+			err := cache.PurgeBlock(bc.headerHashes[i])
+			if err != nil {
+				return fmt.Errorf("error while removing block %d: %w", i, err)
+			}
+		}
+		cache.Store.Put(resetStageKey, []byte{stateResetBit | byte(staleBlocksRemoved)})
+
+		bc.log.Info("blocks, transactions ans AERs are reset", zap.Duration("duration", time.Since(p)))
```

Comment:
Persist before this line?

Reply:
Sure, I might have missed it by accident.

Reply:
Fixed.